### PR TITLE
BigInt typechecking

### LINKED ISSUE #6639: add support for new type: bigint
```javascript
(1n: bigint)
```

https://flow.org/try/#0BQRgdgXABARglgczmALgSiA

### PR DESCRIPTION
**Pull request by goodmind**
Fixes #6639 
Library definitions #7534 (BigInt builtin and typed arrays)

Probably signature verification should be rewritten too, but I don't know what it does.
There maybe missing cases in something like refinements, assignment operators (especially since they seem untyped)

Adds `sketchy-null-bigint` lint rule and supports `bigint` type in `sketchy-number` with different lint message

### CODE REVIEW

File: src/typing/statement.ml [hunk 1]
```diff
@@ -4483,9 +4501,43 @@ and binary cx loc { Ast.Expression.Binary.operator; left; right } =
       let reason = mk_reason (RCustom "arithmetic operation") loc in
       let (_, t1), _ as left = expression cx left in
       let (_, t2), _ as right = expression cx right in
-      Flow.flow cx (t1, AssertArithmeticOperandT reason);
-      Flow.flow cx (t2, AssertArithmeticOperandT reason);
-      NumT.at loc |> with_trust literal_trust, { operator; left; right; }
+      let tout = match (t1, t2) with
```
Comment (goodmind) [inline src/typing/statement.ml]: Probably should be rewritten like `flow_addition`

**Comment by goodmind** [comment]
Not sure why tests fail

**Comment by mvitousek** [comment]
Is there a reason that the names "bigint" and "bignum" are inconsistently used? I'd personally rather align everything with "bigint" since (as far as I understand it) that's the actual Flow type being provided.

**Comment by goodmind** [comment]
@mvitousek just arbitrary, BigNum in typechecker, BigInt in AST

**Comment by goodmind** [comment]
@mvitousek is this all that bothers you?

**Comment by goodmind** [comment]
@mvitousek I renamed all of them to `BigInt`

**Comment by goodmind** [comment]
SO many conflicting files already...

File: src/typing/type.ml [hunk 2]
```diff
@@ -723,10 +770,12 @@ module rec TypeTerm : sig
     | SingletonBoolP of ALoc.t * bool (* true or false *)
     | SingletonStrP of ALoc.t * bool * string (* string literal *)
     | SingletonNumP of ALoc.t * bool * number_literal
+    | SingletonBigIntP of ALoc.t * bool * bigint_literal
 
     | BoolP (* boolean *)
     | FunP (* function *)
     | NumP (* number *)
+    | BIgIntP (* bigint *)
```
Comment (chicoxyzzy) [inline src/typing/type.ml:778]:
```suggestion
    | BigIntP (* bigint *)
```

File: src/typing/debug_js.ml [hunk 3]
```diff
@@ -22,13 +22,15 @@ let string_of_pred_ctor = function
   | BoolP -> "BoolP"
   | StrP -> "StrP"
   | NumP -> "NumP"
+  | BIgIntP -> "BIgIntP"
```
Comment (chicoxyzzy) [inline src/typing/debug_js.ml:25]:
```suggestion
  | BigIntP -> "BigIntP"
```

File: src/typing/debug_js.ml [hunk 4]
```diff
@@ -1388,6 +1400,7 @@ and json_of_pred_impl json_cx p = Hh_json.(
   | StrP
   | SymbolP
   | NumP
+  | BIgIntP
```
Comment (chicoxyzzy) [inline src/typing/debug_js.ml:1403]:
```suggestion
  | BigIntP
```

File: src/typing/flow_js.ml [hunk 5]
```diff
@@ -9142,6 +9314,16 @@ and predicate cx trace t l p = match p with
   | NotP NumP ->
     rec_flow_t cx trace (Type_filter.not_number l, t)
 
+  (***********************)
+  (* typeof _ ~ "bigint" *)
+  (***********************)
+
+  | BIgIntP ->
```
Comment (chicoxyzzy) [inline src/typing/flow_js.ml:9321]:
```suggestion
  | BigIntP ->
```

File: src/typing/flow_js.ml [hunk 6]
```diff
@@ -9142,6 +9314,16 @@ and predicate cx trace t l p = match p with
...
+    rec_flow_t cx trace (Type_filter.bigint l, t)
+
+  | NotP BIgIntP ->
```
Comment (chicoxyzzy) [inline src/typing/flow_js.ml:9324]:
```suggestion
  | NotP BigIntP ->
```

File: src/typing/statement.ml [hunk 7]
```diff
@@ -5537,6 +5649,7 @@ and predicates_of_condition cx e = Ast.(Expression.(
         | "boolean" -> Some BoolP
         | "function" -> Some FunP
         | "number" -> Some NumP
+        | "bigint" -> Some BIgIntP
```
Comment (chicoxyzzy) [inline src/typing/statement.ml:5652]:
```suggestion
        | "bigint" -> Some BigIntP
```

File: newtests/bigint/test.js [hunk 8]
```diff
@@ -201,156 +201,7 @@ export default suite(({ addFile, addFiles, addCode }) => [
       const valid_neg_small = -100n;
       type ValidSmall = 100n;
       type ValidNegSmall = -1n;
```
Comment (chicoxyzzy) [inline newtests/bigint/test.js:203]: Tests for BigInts with numeric separators are missing here. Parser already support them.

**Comment by gkz** [comment]
BigInt is now supported as of 0.194

**Pull request closed without merging**